primer/test/Tests/EvalFull.hs

```diff
@@ -1364,6 +1364,12 @@ unit_prim_partial_map =
             `ann` (tcon tList `tapp` tcon tChar)
             <*> pure (M.singleton mapName mapDef)
    in do
+        -- TODO/REVIEW:
+        -- note that ww/map is somewhat pointless when push down lets -- maybe we should push lets inside letrecs (maybe just inside type?)?
+        --     nb, get something like @let α=Char, β=Char, f=toUpper in (letrec go : List α -> List β; go = λxs.RHS in go)
+        --     and then in two steps (expand @go@, push stack of let+letrec)
+        --        @λxs. let α=Char, β=Char, f=toUpper in (letrec go : List α -> List β; go = λxs.RHS in RHS : List α -> List β)
+        --     we carry around the subst for α,β and f, using α,β inside annotation and f in RHS each time expand the letrec
```

Comment:
> should I elaborate?

Yes please! I'm struggling to see how the first expression becomes the second ("and then in two steps").

Reply:
To elaborate on the above:

Recall that the w/w map is
```
map :: ∀a. ∀b. (a -> b) -> List a -> List b
map = Λa. Λb. λf.
    letrec go :: List a -> List b
           go = λxs. case xs of
                       Nil -> Nil
                       Cons y ys -> Cons (f y) (go ys)
    in go
```
Now consider the application `map Char Char toUpper "abc"`, we get via inlining the global and one BETA step
```
((let a = Char in Λb. λf.
    letrec go :: ...
           go = ...
    in go)
 :: let a = Char in ∀b.(a -> b) -> List a -> List b) Char toUpper "abc"
```
and then we fully reduce the type annotation, before looking at the term (yes this is a bit odd, see my other comment)
```
((let a = Char in Λb. λf.
    letrec go :: ...
           go = ...
    in go)
 :: ∀b.(Char -> b) -> List Char -> List b) Char toUpper "abc"
```
Now one PushLet gives
```
((Λb. let a = Char in λf.
    letrec go :: ...
           go = ...
    in go)
 :: ∀b.(Char -> b) -> List Char -> List b) Char toUpper "abc"
```
which is very similar to the above. Another BETA and reduction of the annotation gives
```
((let b = Char in let a = Char in λf.
    letrec go :: ...
           go = ...
    in go)
 :: (Char -> Char) -> List Char -> List Char) toUpper "abc"
```
Now push the lets again
```
((λf. let b = Char in let a = Char in
    letrec go :: ...
           go = ...
    in go)
 :: (Char -> Char) -> List Char -> List Char) toUpper "abc"
```
And a Beta step
```
((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: ...
           go = ...
    in go)
 :: List Char -> List Char) "abc"
```
Now (and this is the first step of what I above mentioned as "in two steps") we expand the `go`
```
((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: ...
           go = λxs. ...
    in λxs. ...)
 :: List Char -> List Char) "abc"
```
and push the lets
```
((λxs. let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: List a -> List b
           go = λxs. ... Cons (f y) (go ys)
    in case ...)
 :: List Char -> List Char) "abc"
```
Note that this will now reduce by Beta-reducing the `λxs`, and we will keep the stack of lets `f,b,a` around because they are used in the annotation and body of the letrec.
Before we changed to pushing-down-lets, we would have inlined `f,b,a`, giving `letrec go :: List Char -> List Char; go = λxs. case xs of Nil -> Nil ; Cons y ys -> Cons (toUpper y) (go ys)`, which is smaller and faster.

<details>
<summary>Some selected steps in the rest of the reduction, eliding some annotations:</summary>
```
let xs = "abc" in let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: List a -> List b
           go = λxs. ... Cons (f y) (go ys)
    in case ...
```
```
case (let xs = "abc" in xs) of
  Nil -> Nil
  Cons y ys -> let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
               letrec go :: List a -> List b
                      go = λxs. case xs of
                                  Nil -> Nil
                                  Cons y ys -> Cons (f y) (go ys)
               in Cons (f y) (go ys)
```
```
case "abc" of
  Nil -> Nil
  Cons y ys -> let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
               letrec go :: List a -> List b
                      go = λxs. case xs of
                                  Nil -> Nil
                                  Cons y ys -> Cons (f y) (go ys)
               in Cons (f y) (go ys)
```
```
let y='a' in let ys="bc" in
let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
letrec go :: List a -> List b
       go = λxs. case xs of
                    Nil -> Nil
                    Cons y ys -> Cons (f y) (go ys)
in Cons (f y) (go ys)
```
```
Cons (let y='a' in let f = toUpper in f y)
     (let ys="bc" in
      let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
      letrec go :: List a -> List b
             go = λxs. case xs of
                         Nil -> Nil
                         Cons y ys -> Cons (f y) (go ys)
      in go ys)
```
```
Cons 'A'
     ((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
       letrec go :: List a -> List b
              go = λxs. case xs of
                          Nil -> Nil
                          Cons y ys -> Cons (f y) (go ys)
       in go)
      (let ys="bc" in ys))
```
```
Cons 'A'
  (Cons 'B'
     ((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
       letrec go :: List a -> List b
              go = λxs. case xs of
                          Nil -> Nil
                          Cons y ys -> Cons (f y) (go ys)
       in go)
      (let ys="c" in ys)))
```
```
Cons 'A'
  (Cons 'B'
    (Cons 'C'
     ((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
       letrec go :: List a -> List b
              go = λxs. case xs of
                          Nil -> Nil
                          Cons y ys -> Cons (f y) (go ys)
       in go)
      (let ys=Nil in ys))))
```
```
Cons 'A'
  (Cons 'B'
    (Cons 'C'
      (case (let xs = Nil in xs) of
         Nil -> Nil
         Cons y ys -> let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
                      letrec go :: List a -> List b
                             go = λxs. case xs of
                                         Nil -> Nil
                                         Cons y ys -> Cons (f y) (go ys)
                      in Cons (f y) (go ys))))
```
```
Cons 'A' (Cons 'B' (Cons 'C' Nil))
```
</details>

Reply:
TODO: I will write a note about this